### PR TITLE
fix(virtualRepeat): Memory leak

### LINKED ISSUE #8055: mdVirtualRepeat: Memory leak
Blocks created in [VirtualRepeatController](https://github.com/angular/material/blob/v1.1.0-rc3/src/components/virtualRepeat/virtual-repeater.js#L762) are never deleted.

### PR DESCRIPTION
**Pull request by neko1235**
Delete all created elements when the scope is destroyed.

Closes #8055

### CODE REVIEW

**Comment by neko1235** [comment]
@jelbourn Please take a look.

File: src/components/virtualRepeat/virtual-repeater.js [hunk 1]
```diff
@@ -501,6 +501,12 @@ function VirtualRepeatController($scope, $element, $attrs, $browser, $document,
   this.blocks = {};
   /** @type {Array<!VirtualRepeatController.Block>} A pool of presently unused blocks. */
   this.pooledBlocks = [];
+
+  $scope.$on('$destroy', angular.bind(this, function cleanup() {
```
Comment (ThomasBurleson) [inline src/components/virtualRepeat/virtual-repeater.js]: Can you provide a unit test to validate that this works?

**Comment by ThomasBurleson** [comment]
👍

File: src/components/virtualRepeat/virtual-repeater.js [hunk 2]
```diff
@@ -501,6 +501,12 @@ function VirtualRepeatController($scope, $element, $attrs, $browser, $document,
...
+    angular.forEach(this.blocks, removeBlock);
+    angular.forEach(this.pooledBlocks, removeBlock);
+  }));
+  function removeBlock(block) { block.element.remove(); }
```
Comment (jelbourn) [inline src/components/virtualRepeat/virtual-repeater.js]: `element.remove()` unfortunately does not work in IE11. It needs to be
```
if (element.parentNode) {
  element.parentNode.removeChild(element);
}
```

Comment (jelbourn) [inline src/components/virtualRepeat/virtual-repeater.js]: I would also make `removeBlock` a method on the controller's prototype.

File: src/components/virtualRepeat/virtual-repeater.js (hunk 1 above)
Comment (jelbourn) [inline src/components/virtualRepeat/virtual-repeater.js]: Definitely needs a unit test to make ensure that this doesn't regress in the future.

**Comment by jelbourn** [comment]
cc @kseamon

File: src/components/virtualRepeat/virtual-repeater.js (hunk 2 above)
Comment (neko1235) [inline src/components/virtualRepeat/virtual-repeater.js]: > `element.remove()` unfortunately does not work in IE11.

Really? Please note that this is `jqLite.remove()`.

Comment (neko1235) [inline src/components/virtualRepeat/virtual-repeater.js]: How about making `cleanup()` a method on the controller's prototype?

Comment (jelbourn) [inline src/components/virtualRepeat/virtual-repeater.js]: Ah, I didn't realize it was a wrapped element. `remove` is fine, then.

`cleanupBlocks()`  ? Just `cleanup` by itself isn't descriptive.

**Comment by neko1235** [comment]
Please t

**Comment by neko1235** [comment]
Please take a look.

File: src/components/virtualRepeat/virtual-repeater.js [hunk 3]
```diff
@@ -536,6 +538,17 @@ VirtualRepeatController.prototype.link_ =
 };
 
 
+/** @private Cleans up all created blocks (used and unused). */
+VirtualRepeatController.prototype.cleanupBlocks_ = function() {
+  function cleanupBlock(block) {
+    block.element.remove();
+  }
+
+  angular.forEach(this.blocks, cleanupBlock);
```
Comment (kseamon) [inline src/components/virtualRepeat/virtual-repeater.js]: This one is probably not needed (they should be removed along with the parent element). Doesn't hurt much to leave it in, though.

**Comment by kseamon** [comment]
LGTM, overall

**Comment by jelbourn** [comment]
LGTM

**Comment by neko1235** [comment]
Thank you.

File: src/components/virtualRepeat/virtual-repeater.js (hunk 3 above)
Comment (neko1235) [inline src/components/virtualRepeat/virtual-repeater.js]: @kseamon You are correct. Removed.

**Comment by ThomasBurleson** [comment]
👍

**Comment by neko1235** [comment]
Thank you. \o/

**Pull request closed without merging**